jgrapht-core/src/main/java/org/jgrapht/graph/interval/IntervalGraphMapping.java

```diff
@@ -0,0 +1,354 @@
+     *
+     * @throws NullPointerException if the specified edge factory is <code>
+     * null</code>.
+     */
+    private IntervalGraphMapping(ArrayList<V> vertices, Map<Pair<V, V>, E> edges, Supplier<V> vertexSupplier, Supplier<E> edgeSupplier, boolean weighted) {
+        this(new DefaultListenableGraph<>(new DefaultUndirectedGraph<>(vertexSupplier, edgeSupplier, weighted)));
+
+        ArrayList<Interval<T>> vertexIntervals = new ArrayList<>(vertices.size());
+
+        // add sorted vertices
+        for(V v : vertices) {
+            vertexIntervals.add(v.getInterval());
+            intervalMap.put(v.getInterval(), v);
+            graph.addVertex(v);
+        }
+
+        this.intervalStructure = new IntervalTreeStructure<>(vertexIntervals);
+
+        // add all edges
+        for(Pair<V, V> v : edges.keySet()) {
+            graph.addEdge(v.getFirst(), v.getSecond(), edges.get(v));
+        }
+    }
+
+    /**
+     * Getter for the underlying graph.
+     *
+     * @return the underlying graph
+     */
+    public Graph<V, E> getGraph() {
+        return graph;
+    }
+
+    /**
+     * returns interval graph representation if one exists, otherwise null
+     *
+     * @param graph the graph to check
+     * @param <E> the edge type
+     * @param <VertexType> the internal vertex type
+     * @return interval graph representation if one exists, otherwise null
+     */
+    public static <E, VertexType> IntervalGraphMapping<IntervalVertexInterface<VertexType, Integer>, E, VertexType, Integer> asIntervalGraphMapping(Graph<VertexType, E> graph) {
+
+        // initialize IntervalGraphRecognizer
+        IntervalGraphRecognizer<VertexType, E> intervalGraphRecognizer = new IntervalGraphRecognizer<>(graph);
+
+        // execute IntervalGraphRecognizer
+        if(!intervalGraphRecognizer.isIntervalGraph()) {
+            return null;
+        }
+
+        // get necessary data structures from IntervalGraphRecognizer
+        ArrayList<Interval<Integer>> sortedIntervalList = intervalGraphRecognizer.getIntervalsSortedByStartingPoint();
+        Map<Interval<Integer>, VertexType> intervalVertexMap = intervalGraphRecognizer.getIntervalToVertexMap();
+        Map<VertexType, IntervalVertexInterface<VertexType, Integer>> vertexIntervalMap = intervalGraphRecognizer.getVertexToIntervalMap();
+
+        //initialize ArrayList for vertices for quick iteration
+        ArrayList<IntervalVertexInterface<VertexType, Integer>> vertices = new ArrayList<>(sortedIntervalList.size());
+
+        // add vertices to the vertex list
+        for(int i = 0; i < sortedIntervalList.size(); ++i) {
+            vertices.add(i, vertexIntervalMap.get(intervalVertexMap.get(sortedIntervalList.get(i))));
+        }
+
+        // initialze a map from vertex pairs to all edges
+        Map<Pair<IntervalVertexInterface<VertexType, Integer>, IntervalVertexInterface<VertexType, Integer>>, E> edges = new LinkedHashMap<>();
```

Comment:
Why do you use a *Linked*HashMap?

Reply:
good question, a hashmap is ok as well because I want to iterate over the keys...

Reply:
fixed